src/Nethermind/Nethermind.Merge.Plugin/Synchronization/PeerRefresher.cs

```diff
@@ -17,61 +17,202 @@
-            if (timePassed > _minRefreshDelay)
-            {
-                Refresh(blockHash);
-            }
-            else if (!_refreshTimer.Enabled)
-            {
-                _refreshTimer.Interval = _minRefreshDelay - timePassed;
-                _refreshTimer.Start();
-            }
+            Refresh(headBlockhash, headParentBlockhash, finalizedBlockhash);
+        }
+        else if (!_refreshTimer.Enabled)
+        {
+            _refreshTimer.Interval = _minRefreshDelay - timePassed;
+            _refreshTimer.Start();
         }
     }
 
     private void TimerOnElapsed(object? sender, EventArgs e)
     {
-        Refresh(_lastHash);
+        Refresh(_lastBlockhashes.Item1, _lastBlockhashes.Item2, _lastBlockhashes.Item3);
     }
     
-    private void Refresh(Keccak blockHash)
+    private void Refresh(Keccak headBlockhash, Keccak headParentBlockhash, Keccak finalizedBlockhash)
     {
         _lastRefresh = DateTime.Now;
         foreach (PeerInfo peer in _syncPeerPool.AllPeers)
         {
-            _syncPeerPool.RefreshTotalDifficulty(peer.SyncPeer, blockHash);
+            StartPeerRefreshTask(peer.SyncPeer, headBlockhash, headParentBlockhash, finalizedBlockhash);
         }
     }
 
+    private async Task StartPeerRefreshTask(
+        ISyncPeer syncPeer,
+        Keccak headBlockhash,
+        Keccak headParentBlockhash,
+        Keccak finalizedBlockhash
+    )
+    {
+        CancellationTokenSource delaySource = new();
+        Task delayTask = Task.Delay(RefreshTimeout, delaySource.Token);
+        try
+        {
+            await RefreshPeerForFcu(syncPeer, headBlockhash, headParentBlockhash, finalizedBlockhash, delayTask, delaySource.Token);
+        }
+        catch (Exception exception)
+        {
+            if (_logger.IsError) _logger.Error($"Exception in peer refresh. This is unexpected. {syncPeer}", exception);
+        }
+        finally
+        {
+            delaySource.Cancel();
+        }
+    }
+
+    internal async Task RefreshPeerForFcu(
+        ISyncPeer syncPeer,
+        Keccak headBlockhash,
+        Keccak headParentBlockhash,
+        Keccak finalizedBlockhash,
+        Task? delayTask,
+        CancellationToken token
+    ) {
+        
+        // headBlockhash is obtained together with headParentBlockhash
+        Task<BlockHeader[]> getHeadParentHeaderTask = syncPeer.GetBlockHeaders(headParentBlockhash, 2, 0, token);
+        Task<BlockHeader?> getFinalizedHeaderTask = finalizedBlockhash == Keccak.Zero 
+            ? Task.FromResult<BlockHeader?>(null)
+            : syncPeer.GetHeadBlockHeader(finalizedBlockhash, token);
+
+        Task getHeaderTask = Task.WhenAll(getFinalizedHeaderTask, getFinalizedHeaderTask);
+        Task firstToComplete = await Task.WhenAny(getHeaderTask, delayTask);
+
+        if (firstToComplete == delayTask)
+        {
+            _syncPeerPool.ReportRefreshFailed(syncPeer, "timeout");
+            return;
+        }
```

Comment:
If we timeout and didn't get any response (WhenAll is for all responses), then maybe we want to disconnect?